File: backend/src/main/java/com/insteip/backend/service/interfaces/CertificadoService.java
package com.insteip.backend.service.interfaces;

import com.insteip.backend.entity.Certificado;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface CertificadoService {
    Certificado generarCertificado(Long usuarioId, Long cursoId);
    Certificado obtenerCertificado(Long id);
    Certificado validarCertificado(String codigo);
    Page<com.insteip.backend.dto.CertificadoResponseDTO> listarCertificados(Pageable pageable, String search);
}

File: frontend/src/app/core/services/certificado.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface CertificadoResponse {
  id: number;
  usuarioId: number;
  cursoId: number;
  codigo: string;
  archivoPdf: string;
  urlValidacion: string;
  numeroRegistro: string;
  fechaEmision: string;
}

@Injectable({
  providedIn: 'root'
})
export class CertificadoService {
  private http = inject(HttpClient);
  private apiUrl = environment.apiUrl + '/certificados';

  generarCertificado(cursoId: number): Observable<CertificadoResponse> {
    return this.http.post<CertificadoResponse>(`${this.apiUrl}/generar/${cursoId}`, null);
  }

  validarCertificado(codigo: string): Observable<CertificadoResponse> {
    return this.http.get<CertificadoResponse>(`${this.apiUrl}/validar/${codigo}`);
  }

  listarCertificados(page = 0, size = 5, search = '', sort = 'fechaEmision,desc'): Observable<any> {
    return this.http.get<any>(this.apiUrl, {
      params: {
        page: String(page),
        size: String(size),
        ...(search.trim() ? { search: search.trim() } : {}),
        sort
      }
    });
  }
}
